Fix invalid or incorrect UTF-8 characters

rmunn opened the pull request:
Problems fixed in this commit:

  - Some curly-quotes encoded in CP-1252. Replaced with ASCII quotes.
  - Some cases of – (U+2013 EN DASH) used in comments where a hyphen was clearly intended (e.g., in the number -1). Replaced with hyphen.
  - Many © characters (and one ® character) encoded in CP-1252 instead of UTF-8. Replaced with correct UTF-8 characters.
  - Two cases of ö encoded in CP-1252 instead of UTF-8. Fixed.

Problems NOT fixed in this commit:

  - [Microsoft.PowerShell.Utility.psm1](https://github.com/PowerShell/PowerShell/blob/32a8601f926b164bbdf74ad1b6c4c8ef927fe1a8/src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1#L377) file contains some invalid UTF-8 bytes in comments on line 377 and lines 390-392. Since those are meant to represent CP-1252 encoded output on the Windows command prompt, it's less clear what to do here. Left them alone for now.
  - Many files have a UTF-8 BOM that's no longer the first character of the file. Usually this is because a "Copyright (C) Microsoft.  All rights reserved." comment has been automatically added at the top of the file by some script which didn't properly handle the existence of a BOM character. One example of this problem can be seen in [DefinitionCache.cs](https://github.com/PowerShell/PowerShell/blob/32a8601f926b164bbdf74ad1b6c4c8ef927fe1a8/src/Microsoft.PowerShell.Workflow.ServiceCore/ServiceCore/WorkflowCore/DefinitionCache.cs#L4), where line 4 starts with the BOM that should be at the start of line 1. This is a different issue and should be tracked (and fixed) separately.

This should mostly fix #2851, and any remaining Unicode problems (such as the BOM issue mentioned above) should probably be tracked in a new issue.

rmunn committed:
Fix Unicode in license_thirdparty_proprietary.txt

This file contained many occurrences of U+F0A7, a character from the
designated Private Use Area, so its meaning is not defined and fonts
can assign any symbol to that character. Since the person who originally
wrote this file works at Microsoft, it seems reasonable to assume that
he was using the Wingdings font, which has a bullet character at U+F0A7.
Thus, replacing these characters with the proper Unicode bullet (U+2022)
seems the right way to fix this.

rmunn committed:
Fix invalid or incorrect UTF-8 characters

Problems fixed in this commit:

  - Some curly-quotes encoded in CP-1252. Replaced with ASCII quotes.
  - Some cases of – (U+2013 EN DASH) used in comments where a hyphen was
    clearly intended (e.g., in the number -1). Replaced with hyphen.
  - Many © characters (and one ® character) encoded in CP-1252 instead
    of UTF-8. Replaced with correct UTF-8 characters.
  - Two cases of ö encoded in CP-1252 instead of UTF-8. Fixed.

Problems NOT fixed in this commit:

  - Microsoft.PowerShell.Utility.psm1 file contains some invalid UTF-8
    bytes in comments on line 377 and lines 390-392. Since those are
    meant to represent CP-1252 encoded output on the Windows command
    prompt, it's less clear what to do here. Left them alone for now.
  - Many files have a UTF-8 BOM that's no longer the first character
    of the file. Usually this is because a "Copyright (C) Microsoft.
    All rights reserved." comment has been automatically added at the
    top of the file by some script which didn't properly handle the
    existence of a BOM character. One example of this problem can be
    seen in DefinitionCache.cs. This is a different issue and should
    be tracked (and fixed) separately.

This should mostly fix #2851, and any remaining Unicode problems (such
as the BOM issue above) should probably be tracked in a new issue.

msftclas commented:
Hi __@rmunn__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

rmunn commented:
Note that https://github.com/PowerShell/PowerShell/pull/2857 is a subset of these changes. I created that PR first, then decided to go ahead and fix the rest of the Unicode encoding problems I had found. So if this PR is accepted, https://github.com/PowerShell/PowerShell/pull/2857 will be moot.

iSazonov commented:
1. I think we should fix Microsoft.PowerShell.Utility.psm1 here too because the problem is in _comments_ , not in real data.
2. We should add tests. Perhaps we can add CI encoding checks as requested by @vors in #1559

rmunn commented:
I've signed the CLA, but the MSBOT hasn't updated the PR yet. Do I need to email the signed PDF back to Microsoft or something?

SteveL-MSFT commented on src/System.Management.Automation/engine/InternalCommands.cs:
```diff
@@ -295,12 +295,12 @@ protected override void ProcessRecord()
                                 {
                                     // we write null out because:
-                                    // PS C:\> �$null | ForEach-object {$_.aa} | ForEach-Object {$_ + 3}�
+                                    // PS C:\> "$null | ForEach-object {$_.aa} | ForEach-Object {$_ + 3}"
```
adding quotes here changes the meaning of that comment line from a powershell pipeline to a string

rmunn replied:
The original file before my fixes contained a `0x93` byte at the start of the comment line, and a `0x94` byte at the end of the comment line. In CP-1252, those are opening and closing curly quotes, so I replaced them with straight quotes since that was the closest thing to the original intent of whoever wrote that comment.

But since that comment's original intent was apparently *incorrect*, I'll be happy to remove the quotes.

SteveL-MSFT commented on src/System.Management.Automation/engine/InternalCommands.cs:
```diff
@@ -295,12 +295,12 @@ protected override void ProcessRecord()
                                     // 3
                                     // so we also want
-                                    // PS C:\> �$null | ForEach-object aa | ForEach-Object {$_ + 3}�
+                                    // PS C:\> "$null | ForEach-object aa | ForEach-Object {$_ + 3}"
```
adding quotes here changes the meaning of that comment line from a powershell pipeline to a string

rmunn replied:
Same comment as before: that was the intent of the original author, given the CP-1252-encoded curly quotes that I found. But I'll be happy to remove the quotes if they're actually incorrect.

SteveL-MSFT replied:
Same, quotes shouldn't have been there in the first place.

SteveL-MSFT commented on src/System.Management.Automation/engine/Modules/NewModuleManifestCommand.cs:
```diff
@@ -837,7 +837,7 @@ private string ManifestComment(string insert, StreamWriter streamWriter)
         protected override void EndProcessing()
         {
-            // Win8: 264471 - Error message for New-ModuleManifest �ProcessorArchitecture is obsolete.
+            // Win8: 264471 - Error message for New-ModuleManifest - ProcessorArchitecture is obsolete.
```
ProcessorArchitecture is a parameter, so it should be written as:

// Win8: 264471 - Error message for New-ModuleManifest -ProcessorArchitecture is obsolete.

rmunn replied:
It was an en-dash character, so I figured it was meant to be a dash. If it's supposed to be a single-hyphen prefix, then that's what I'll put in there.

SteveL-MSFT replied:
Should be single-hyphen indicating it's a parameter passed to new-modulemanifest cmdlet

SteveL-MSFT commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -572,7 +572,7 @@ private string FindHelpFile(CmdletInfo cmdletInfo)
 
                 // When cmdlet metadata is generated for the 'HelpFile' field, we use the name assembly and we append '-Help.xml' to it.
-                // Because of this, if the cmdlet is part of an N�gen assembly, then 'HelpFile' field will be pointing to a help file which does not exist.
+                // Because of this, if the cmdlet is part of an N'gen assembly, then 'HelpFile' field will be pointing to a help file which does not exist.
```
Ngen refers to Ngen.exe which produces native code from managed code, so the apostrophe doesn't make sense

rmunn replied:
The original comment had a `0x92` byte, which is a curly apostrophe in CP-1252. I did wonder about whether it made sense, but I wanted to stay as close to the original as possible. I'll be happy to fix it to what it *should* be, though.

SteveL-MSFT replied:
Should be just Ngen

msftgits commented:
Hi, I am closing and re-opening this PR to bump the CLA bot. Sorry for the inconvenience!

msftclas commented:
Hi __@rmunn__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

rmunn committed:
Fix comments whose punctuation was incorrect

This implements all the changes requested by Steve Lee.

SteveL-MSFT reviewed (approved):
LGTM

lzybkr commented on src/Modules/Shared/Microsoft.PowerShell.Archive/Microsoft.PowerShell.Archive.psd1:
```diff
@@ -2,7 +2,7 @@
 Author="Microsoft Corporation"
 CompanyName="Microsoft Corporation"
-Copyright="� Microsoft Corporation. All rights reserved."
+Copyright="© Microsoft Corporation. All rights reserved."
```
Some Windows tools will not handle this file correctly because there is no BOM.  For example, gvim shows 2 characters for the copyright symbol, I assume because it guesses the file is Latin-1, not UTF8.

rmunn committed:
Add UTF-8 BOM to files that will need it

Although the BOM "is neither required nor recommended" for UTF-8 files
by the Unicode standard, it is still necessary for some tools to process
these files correctly now that they contain the © (copyright) character.
In the absence of a BOM, some tools still default to Latin-1 instead of
UTF-8; until such tools are changed to default to a Unicode encoding,
putting a BOM in front of UTF-8 files remains an unfortunate necessity.

rmunn committed:
Add UTF-8 BOM to two more files that will need it

iSazonov commented:
Why we use unwanted BOM instead of simply replace one © (copyright) character? Now 1457 file contain "Copyright (C)".

lzybkr commented:
I'm closing this PR as we're going a different route.
First, some of these files will be removed when we remove Windows (full) PowerShell from this repo.
For other files, I think we want pure ASCII and remove any BOMs - necessary or not - which means we'll replace some of the characters, e.g. replace copyright with (c).
